DAOS-7843 daos: use dfs_sys API in fs copy

dsikich opened the pull request:
* fix return code errors in fault injection tests
    * use macros to print all error codes with standard format
    * fix when error code  was not being properly set
    * return DAOS error codes from higher level functions
    * use dfs_sys API to replace old dfs wrappers

Test-tag: pr fs_copy cont_clone

Signed-off-by: Danielle Sikich <danielle.sikich@intel.com>

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2096,143 +2000,59 @@ file_mkdir(struct cmd_args_s *ap, struct file_dfs *file_dfs,
 file_readdir(struct cmd_args_s *ap, struct file_dfs *file_dfs, DIR *dirp)
 {
 	struct dirent *entry = NULL;
+	int rc		     = 0;			
```
(style)  trailing whitespace

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2922,64 +2659,67 @@ dm_connect(struct cmd_args_s *ap,
-			dattr.da_props = props;
-			rc = duns_create_path(ca->dst_poh, path,
-					      &dattr);
+			if (ca->cont_layout != DAOS_PROP_CO_LAYOUT_POSIX) 
```
(style)  trailing whitespace

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2922,64 +2659,67 @@ dm_connect(struct cmd_args_s *ap,
-			rc = dfs_mount(ca->dst_poh, ca->dst_coh, O_RDWR,
-				       &dst_file_dfs->dfs);
+			rc = dfs_sys_mount(ca->dst_poh, ca->dst_coh, O_RDWR, DFS_SYS_NO_LOCK,
+		       			   &dst_file_dfs->dfs_sys);
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -3004,12 +2743,11 @@ dm_connect(struct cmd_args_s *ap,
-				DP_UUID(ca->src_p_uuid),
-				DP_RC(rc2));
+			DH_PERROR_DER(ap, rc2, "Failed to disconnect from source pool "DF_UUIDF,
+			      	      DP_UUID(ca->dst_p_uuid));
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -3701,101 +3407,88 @@ cont_clone_hdlr(struct cmd_args_s *ap)
-				       &epoch, NULL,
-			       DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT, NULL);
+	rc = daos_cont_create_snap_opt(ca.src_coh, &epoch, NULL,
+       				       DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT, NULL);
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/1/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/1/execution/node/100/log

daosbuild1 commented:
Test stage Build on CentOS 7 debug completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/1/execution/node/409/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/2/execution/node/422/log

daosbuild1 commented:
Test stage Build on CentOS 7 debug completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/2/execution/node/444/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/2/execution/node/389/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented on src/client/dfs/dfs_sys.c:
```diff
@@ -213,7 +213,7 @@ hash_lookup(dfs_sys_t *dfs_sys, struct sys_path *sys_path)
 	D_STRNDUP(hdl->name, sys_path->dir_name, sys_path->dir_name_len);
 	if (hdl->name == NULL)
-		D_GOTO(free_hdl, ENOMEM);
+		D_GOTO(free_hdl, rc = ENOMEM);
```
FYI: This partially resolves https://jira.hpdd.intel.com/browse/DAOS-8218

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1900,99 +1902,49 @@ parse_filename_dfs(const char *path, char **_obj_name, char **_cont_name)
 	} else {
-		fprintf(ap->errstream, "File type not known: %s type=%d\n",
-			file, file_dfs->type);
+		rc = EINVAL;
+		DH_PERROR_SYS(ap, rc, "File type not known '%s' type=%d", file, file_dfs->type);
```
Maybe this should also return -1 for num_bytes_written. Otherwise the caller doesn't know that this call failed

dsikich replied:
@daltonbohning I thought dfs_sys_write returned -1 on failure? For the POSIX case write will return -1 on error.

daltonbohning replied:
But this is the case where an incorrect file type was passed. -1 is not returned in this case.

dsikich replied:
@daltonbohning yeah, but I think maybe in that case I should check for EINVAL after the functional is called? Since it is different than read/write actually failing.

daltonbohning replied:
@dsikich This is the full function: https://github.com/daos-stack/daos/blob/b447b409375e2a8f193ccdc1868134c3a564160b/src/utils/daos_hdlr.c#L1918-L1946
The problem is that num_bytes_written is returned, not rc. So rc can't be checked. This is part of why I suggested (different comment) changing these functions to have signatures similar to the dfs / dfs_sys signatures, where rc is always returned and something like num_bytes_written is passed/returned by reference.

dsikich replied:
@daltonbohning hmm yeah I think for the wrappers that would be better. I will change these to return rc and fill in num_bytes_written or bytes read.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2096,143 +2000,59 @@ file_mkdir(struct cmd_args_s *ap, struct file_dfs *file_dfs,
-		rc = mkdir_dfs(ap, file_dfs, dir, mode);
+		rc = dfs_sys_mkdir(file_dfs->dfs_sys, dir, *mode, 0);
 		if (rc != 0) {
 			/* mkdir_dfs already prints error */
```
This isn't true about dfs_sys_mkdir, which doesn't print any errors

dsikich replied:
@daltonbohning yeah sorry this was an old comment that got left here.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2096,143 +2000,59 @@ file_mkdir(struct cmd_args_s *ap, struct file_dfs *file_dfs,
+			rc = errno;
 	} else if (file_dfs->type == DAOS) {
-		dirp = opendir_dfs(ap, file_dfs, dir);
+		rc = dfs_sys_opendir(file_dfs->dfs_sys, dir, 0, &dirp);
+		if (rc != 0)
+			dirp = NULL;
```
NULL assignment shouldn't be necessary here, since if the call failed dfs_sys_opendir should/would leave the pointer untouched
```suggestion
		rc = dfs_sys_opendir(file_dfs->dfs_sys, dir, 0, &dirp);
```

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2096,143 +2000,59 @@ file_mkdir(struct cmd_args_s *ap, struct file_dfs *file_dfs,
+			rc = errno;
 	} else if (file_dfs->type == DAOS) {
-		entry = readdir_dfs(ap, file_dfs, dirp);
+		rc = dfs_sys_readdir(file_dfs->dfs_sys, dirp, &entry);
+		if (rc != 0)
+			entry = NULL;
```
Similar to dfs_sys_mkdir, the pointer should be untouched on failure
```suggestion
		rc = dfs_sys_readdir(file_dfs->dfs_sys, dirp, &entry);
```

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2244,83 +2064,45 @@ file_lstat(struct cmd_args_s *ap, struct file_dfs *file_dfs,
-		}
 	} else if (file_dfs->type == DAOS) {
-		rc = stat_dfs(ap, file_dfs, path, buf);
+		rc = dfs_sys_stat(file_dfs->dfs_sys, path, 0, buf);
```
Technically lstat should be:
```suggestion
		rc = dfs_sys_stat(file_dfs->dfs_sys, path, O_NOFOLLOW, buf);
```
Otherwise links are followed by default

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2096,143 +2000,59 @@ file_mkdir(struct cmd_args_s *ap, struct file_dfs *file_dfs,
 	} else {
-		fprintf(ap->errstream, "File type not known, type=%d\n",
-			file_dfs->type);
+		DH_PERROR_SYS(ap, rc, "File type not known type=%d", file_dfs->type);
```
rc would be zero here, but should probably be EINVAL

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2096,143 +2000,59 @@ file_mkdir(struct cmd_args_s *ap, struct file_dfs *file_dfs,
 	} else {
-		fprintf(ap->errstream, "File type not known: %s type=%d\n",
-			dir, file_dfs->type);
+		DH_PERROR_SYS(ap, rc, "File type not known '%s' type=%d", dir, file_dfs->type);
```
rc would be zero here, but should probably be EINVAL

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2244,83 +2064,45 @@ file_lstat(struct cmd_args_s *ap, struct file_dfs *file_dfs,
-		fprintf(ap->errstream,
-			"File type not known, file=%s, type=%d\n",
-			path, file_dfs->type);
+		DH_PERROR_SYS(ap, rc, "File type not known '%s' type=%d", path, file_dfs->type);
```
rc would be zero here, but should probably be EINVAL

dsikich replied:
@daltonbohning thanks, I guess I missed a couple of these.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2244,83 +2064,45 @@ file_lstat(struct cmd_args_s *ap, struct file_dfs *file_dfs,
-		got_size = -1;
-		fprintf(ap->errstream, "File type not known: %s type=%d\n",
-			file, file_dfs->type);
+		DH_PERROR_SYS(ap, rc, "File type not known '%s' type=%d", file, file_dfs->type);
```
rc would be zero here, but should probably be EINVAL

daltonbohning replied:
Maybe this should return -1 for got_size so the caller knows it failed

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2372,51 +2142,13 @@ file_close(struct cmd_args_s *ap, struct file_dfs *file_dfs, const char *file)
 		}
 	} else if (file_dfs->type == DAOS) {
-		rc = close_dfs(ap, file_dfs, file);
+		dfs_sys_close(file_dfs->obj);
```
```suggestion
		rc = dfs_sys_close(file_dfs->obj);
```

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2487,8 +2218,7 @@ fs_copy_file(struct cmd_args_s *ap,
 		ssize_t bytes_read = file_read(ap, src_file_dfs, src_path,
 					       buf, left_to_read);
 		if (bytes_read < 0) {
-			rc = EIO;
-			DH_PERROR_SYS(ap, rc, "read 'failed on '%s'", src_path);
+			rc = daos_errno2der(rc);
```
There is no rc from this function. Maybe these file_* functions should have signatures similar to the dfs_sys versions? I.e. bytes_read is passed/returned by reference and rc is returned from the function.
Then you would have the corresponding rc

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2497,18 +2227,17 @@ fs_copy_file(struct cmd_args_s *ap,
 		bytes_written = file_write(ap, dst_file_dfs, dst_path,
 					   buf, bytes_to_write);
 		if (bytes_written < 0) {
-			rc = EIO;
-			DH_PERROR_SYS(ap, rc, "write failed on '%s'", dst_path);
+			rc = daos_errno2der(rc);
```
Same. There is no rc from this function

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2542,20 +2271,20 @@ fs_copy_dir(struct cmd_args_s *ap,
 	/* begin by opening source directory */
 	src_dir = file_opendir(ap, src_file_dfs, src_path);
 	if (!src_dir) {
-		rc = errno;
-		DH_PERROR_SYS(ap, rc, "Cannot open directory '%s'", src_path);
+		rc = daos_errno2der(rc);
+		DH_PERROR_DER(ap, rc, "Cannot open directory '%s'", src_path);
```
There is no rc from this function

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2810,13 +2539,15 @@ dm_connect(struct cmd_args_s *ap,
-	dfs_attr_t		attr = {0};
 	int			size = 2;
 	uint32_t		dpe_types[size];
-	uint64_t		dpe_vals[size];
+	uint64_t		*dpe_vals = NULL;
```
Why make dpe_types static and dpe_vals dynamic since they are both [size]?

dsikich replied:
@daltonbohning I think the reason I did this is because dpe_types is not modified in dm_get_cont_prop function (the values are set before), but the dpe_vals are set within the function.

dsikich replied:
@daltonbohning I admit I cannot remember exact reason here.

dsikich replied:
@daltonbohning I do remember dpe_vals needs to modify array values in dm_cont_get_prop function whereas dpe_types does not.

dsikich replied:
@daltonbohning yeah it used to be static array. I will change it back to make less confusing.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -3626,30 +3341,29 @@ cont_clone_hdlr(struct cmd_args_s *ap)
-		fprintf(ap->errstream,
-			"failed to parse destination path: %d\n", rc);
-		D_GOTO(out, rc = daos_errno2der(rc));
+		rc = daos_errno2der(rc);
+		DH_PERROR_DER(ap, rc, "Failed to parse destination path");
```
dm_parse_path returns a -DER now, right? So this errno2der shouldn't be needed

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -3670,27 +3383,20 @@ cont_clone_hdlr(struct cmd_args_s *ap)
-				"or provide an existing pool or new UNS path "
-				"of the form:\n\t--dst </$pool> | "
-				"<path/to/uns>\n");
+				"This destination container already exists, please provide a "
```
Grammar nitpick
```suggestion
				"This destination container already exists; please provide a "
```
Or
```suggestion
				"This destination container already exists. Please provide a "
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,14 +2294,21 @@ fs_copy_dir(struct cmd_args_s *ap,
 			break;
-		}
-
+		} 
```
(style)  trailing whitespace

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,14 +2294,21 @@ fs_copy_dir(struct cmd_args_s *ap,
-		}
-
+		} 
+		
```
(style)  trailing whitespace

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/5/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/5/execution/node/100/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1915,34 +1915,32 @@ parse_filename_dfs(const char *path, char **_obj_name, char **_cont_name)
 
 	if (file_dfs->type == POSIX) {
-		num_bytes_written = write(file_dfs->fd, buf, size);
+		write_size = write(file_dfs->fd, buf, *size);
+		*size = write_size;
```
Minor: would this be simpler?
```suggestion
		*size = write(file_dfs->fd, buf, *size);
```

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2067,40 +2070,39 @@ file_lstat(struct cmd_args_s *ap, struct file_dfs *file_dfs,
-		got_size = read_size;
+			file_dfs->offset += (daos_off_t)*size;
 	} else {
 		DH_PERROR_SYS(ap, rc, "File type not known '%s' type=%d", file, file_dfs->type);
```
This still needs to set rc = EINVAL, right?

dsikich replied:
@daltonbohning ah I though I caught all of those again, thanks

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2295,7 +2294,14 @@ fs_copy_dir(struct cmd_args_s *ap,
+		rc = file_readdir(ap, src_file_dfs, src_dir, &entry);
+		if (rc != 0) {
+			rc = daos_errno2der(rc);
+			DH_PERROR_DER(ap, rc, "Cannot read directory");
```
Should this print the directory?
```suggestion
			DH_PERROR_DER(ap, rc, "Cannot read directory '%s'", src_dir);
```

dsikich replied:
@daltonbohning well I can cast it to (char*)? Since the macro will only take that and src_dir is a DIR*

daltonbohning replied:
@dsikich Oops. Didn't realize is was DIR*. Maybe this is okay then

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2628,7 +2357,8 @@ fs_copy_dir(struct cmd_args_s *ap,
 	rc = file_chmod(ap, dst_file_dfs, dst_path, src_stat->st_mode);
 	if (rc != 0) {
-		DH_PERROR_SYS(ap, rc, "updating destination permissions failed on '%s'", dst_path);
+		rc = daos_errno2der(rc);
```
curious why you are changing return codes to DER from errno?

dsikich replied:
@mchaarawi all of the higher level functions just return DER codes. The file_* wrappers themselves still return errno codes but are converted in calling functions. I suppose the idea was just to do it consistently. It could have returned only errno codes from all higher level functions but the thinking is that since it is a user facing DAOS utility it might be better if they all returned DER codes.

dsikich replied:
@mchaarawi all of the higher level functions just return DER codes. The file_* wrappers themselves still return errno codes but are converted in calling functions. I suppose the idea was just to do it consistently. It could have returned only errno codes from all higher level functions but the thinking is that since it is a user facing DAOS utility it might be better if they all returned DER codes.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -53,6 +54,7 @@ struct file_dfs {
 	daos_off_t offset;
 	dfs_obj_t *obj;
 	dfs_t *dfs;
```
if you are switching to the dfs_sys api, is dfs_t *dfs, still needed here?

dsikich replied:
@mchaarawi no, removed it.

mchaarawi reviewed (commented):
I would have rather preferred you split this in 2 PRs
1) to enable fi testing.
2) to switch to sys API.
but not going to request that change now, just for future PRs, incase one needs reverting ig it causes for example daily regressions, etc.

anyway, NLT tests are failing.. 
and you need to push this with daily_regression tag to test more of the dm stuff, no?

daltonbohning commented:
> and you need to push this with daily_regression tag to test more of the dm stuff, no?

This actually tests ALL fs copy  and cont clone tests, including daily_regression and full_regression: `Test-tag: pr fs_copy cont_clone`

dsikich commented:
> I would have rather preferred you split this in 2 PRs
> 
> 1. to enable fi testing.
> 2. to switch to sys API.
>    but not going to request that change now, just for future PRs, incase one needs reverting ig it causes for example daily regressions, etc.
> 
> anyway, NLT tests are failing..
> and you need to push this with daily_regression tag to test more of the dm stuff, no?

@mchaarawi @daltonbohning addressed the comments about datamover tests, so they are all being ran on this PR for fs_copy and cont_clone. My thinking in putting in one PR was mostly because the old dfs_* wrappers were causing fault injection test issues for fs copy, and I didn't want to spend a bunch of time fixing them in one PR to delete them in the next one. I see how breaking them up would make it easier though. I am sort of still learning the NLT output and I fixed all of the daos_hdlr.c file issues, but I guess this PR introduced like 6 more warnings from other files, so I will try to track down those.

dsikich commented:
> > > I would have rather preferred you split this in 2 PRs
> > > 
> > > 1. to enable fi testing.
> > > 2. to switch to sys API.
> > >    but not going to request that change now, just for future PRs, incase one needs reverting ig it causes for example daily regressions, etc.
> > > 
> > > anyway, NLT tests are failing..
> > > and you need to push this with daily_regression tag to test more of the dm stuff, no?
> > 
> > 
> > @mchaarawi @daltonbohning addressed the comments about datamover tests, so they are all being ran on this PR for fs_copy and cont_clone. My thinking in putting in one PR was mostly because the old dfs_* wrappers were causing fault injection test issues for fs copy, and I didn't want to spend a bunch of time fixing them in one PR to delete them in the next one. I see how breaking them up would make it easier though. I am sort of still learning the NLT output and I fixed all of the daos_hdlr.c file issues, but I guess this PR introduced like 6 more warnings from other files, so I will try to track down those.
> 
> just to be clear, the FI test you enabled is causing 37 new warnings. the 6 i tihnk are generic NLT errors for the changes you made. that's my understanding of the result, but i could be wrong

@mchaarawi yeah took me a bit to figure out the output properly since I am new to using it, and I see them now. Actually, one fault injection is causing a cascade of about 23 errors, so I am working on fixing them now.

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1419,8 +1419,8 @@ dfs_cont_create_int(daos_handle_t poh, uuid_t *cuuid, bool uuid_is_set, uuid_t i
-			D_GOTO(err_destroy, rc = daos_der2errno(rc));
-		}
+			D_GOTO(err_close, rc = daos_der2errno(rc));
+                }
```
(style)  code indent should use tabs where possible

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2912,102 +2639,95 @@ dm_connect(struct cmd_args_s *ap,
+	if (daos_handle_is_valid(ca->dst_poh)) {
+		rc2 = daos_pool_disconnect(ca->dst_poh, NULL);
+		if (rc2 != 0)
+			DH_PERROR_DER(ap, rc2, "failed to disconnect from destination pool "DF_UUIDF,
```
(style)  line over 100 characters

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -3033,54 +2753,57 @@ dm_disconnect(struct cmd_args_s *ap,
+	 * code immediately after a failure to umount/close/disconnect then fault injection
+	 * will always report a memory leak. Is it better to immediately return if one of
+	 * these fails? This will cause memory leaks in fault injection tests for fs copy,
+	 * so not sure what is the best thing to do here. 
```
(style)  trailing whitespace

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/8/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/8/execution/node/87/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented on src/client/dfs/dfs.c:
```diff
@@ -3327,10 +3327,11 @@ dfs_release(dfs_obj_t *obj)
-	else
+	} else {
 		D_FREE(obj);
+	}
```
Why add brackets? I think the preferred style is to not add them if not needed

dsikich replied:
@daltonbohning yeah I removed this in the next commit I pushed for check patch. Didn't mean to add these.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2767,56 +2753,57 @@ dm_disconnect(struct cmd_args_s *ap,
+	 * will always report a memory leak. Is it better to immediately return if one of
+	 * these fails? This will cause memory leaks in fault injection tests for fs copy,
+	 * so not sure what is the best thing to do here. 
+	 */
```
Similar to the conversation between @mchaarawi and @ashleypittman. At this time, I'm not sure what the best solution is.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on CentOS 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/10/execution/node/391/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/10/execution/node/397/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/10/execution/node/394/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/10/execution/node/346/log

daosbuild1 commented:
Test stage Build RPM on CentOS 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/11/execution/node/284/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/11/execution/node/277/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6353/11/execution/node/301/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented on src/client/dfs/dfs_sys.c:
```diff
@@ -227,8 +227,7 @@ hash_lookup(dfs_sys_t *dfs_sys, struct sys_path *sys_path)
 	if (rc != 0) {
-		D_ERROR("dfs_lookup() %s failed: %s\n",
-			sys_path->dir_name, strerror(rc));
+		D_DEBUG(DB_TRACE, "dfs_lookup() %s failed: %s\n", sys_path->dir_name, strerror(rc));
```
I'm okay with this (debug probably makes more sense anyway), but how does this related to fault injection? (genuine question)

dsikich replied:
@daltonbohning for some reason it was giving a "warning in strict mode" in the nlt-errors.json file. I am not sure why, but using DEBUG instead removed the error from the file.

ashleypittman replied:
If it's an error then use D_ERROR().

The actual code doing the log check is at https://github.com/daos-stack/daos/blob/master/src/tests/ftest/cart/util/cart_logtest.py#L401-L424 and it's attempting to verify that you're using the DF_RC macro but that only applies to daos error numbers, for system error numbers it's doing a much looser check that the error number itself (12) is printed.  I think in this case changing back the line and modifying cart_lotest.py to accept the strerror() string.

daltonbohning replied:
@ashleypittman This was the only D_ERROR in this file. Everything else is DEBUG. The idea was that most of these functions are pass-through / shorthand for DFS. So, either (1) DFS itself would log appropriate error messages, or (2) the caller of the dfs_sys functions would check the return codes.

daosbuild1 commented:
Test stage Scan CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/12/display/redirect

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/12/display/redirect

daltonbohning reviewed (dismissed):
Assuming NLT passes, LGTM

dsikich commented:
> Assuming NLT passes, LGTM

@daltonbohning @mchaarawi NLT is passing now with fs copy fault injection turned on.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1922,7 +1922,8 @@ file_write(struct cmd_args_s *ap, struct file_dfs *file_dfs,
 	if (file_dfs->type == POSIX) {
 		*size = write(file_dfs->fd, buf, *size);
-		if (*size < 0)
+		/* cast to fix compiler warning about comparison of unsigned type */
+		if ((int)*size < 0)
```
This should be cast to the type returned from write
https://man7.org/linux/man-pages/man2/write.2.html
```suggestion
		/* cast to fix compiler warning about comparison of unsigned type */
		if ((ssize_t)*size < 0)
```

daltonbohning replied:
Actually, daos_size_t is unsigned
```
typedef uint64_t      daos_size_t
```
So I think the ways dfs_sys sets -1 is incorrect, right?

daltonbohning replied:
This is a larger issue than this PR - I'm going to create a ticket with the proper fix

dsikich replied:
@daltonbohning yeah, thanks. I updated the file_read/write calls to use ssize_t

daltonbohning replied:
@dsikich Thanks. LGTM. As we discussed, either way we are going to be limited in some way by ssize_t, since that is what the POSIX functions use. But this is okay for now since we don't pass sizes that would overflow the positive range of ssize_t. Maybe that could change in the future if the chunk size is made configurable by the user, at which time we could re-address this and make changes if necessary.
Also, I created https://jira.hpdd.intel.com/browse/DAOS-8293 for the dfs_sys part of this issue

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2923,102 +2659,95 @@ dm_connect(struct cmd_args_s *ap,
-						      "failed to create destination container");
-					D_GOTO(err_dst_root, rc = daos_errno2der(rc));
+					rc = daos_errno2der(rc);
+					DH_PERROR_DER(ap, rc, "failed to create destination "
+						      "container");
```
Minor, but it would be better to split this after a ,
```suggestion
					DH_PERROR_DER(ap, rc,
								       "failed to create destination container");
```

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2923,102 +2659,95 @@ dm_connect(struct cmd_args_s *ap,
+		rc2 = daos_pool_disconnect(ca->dst_poh, NULL);
+		if (rc2 != 0)
+			DH_PERROR_DER(ap, rc2, "failed to disconnect from destination pool "
+				      ""DF_UUIDF, DP_UUID(ca->dst_p_uuid));
```
```suggestion
				      DF_UUIDF, DP_UUID(ca->dst_p_uuid));
```

ashleypittman reviewed (commented):
I think @mchaarawi is right for now, the fault injection testing (in particular the new fs copy test) is good at flushing out issues (and daos_hdlr.c has many) but the cost in terms of time-to-result of turning it on for everything at the minute is just too high.  I've got some ideas to speed up and parallelise the testing but that's a month or so out.

I think for now it's fine to enable the test for the development process and to use it when working on this code but until I can reduce the burden of it we should avoid enabling it by default, even if all the issues are fixed.

dsikich commented:
> I think @mchaarawi is right for now, the fault injection testing (in particular the new fs copy test) is good at flushing out issues (and daos_hdlr.c has many) but the cost in terms of time-to-result of turning it on for everything at the minute is just too high. I've got some ideas to speed up and parallelise the testing but that's a month or so out.
> 
> I think for now it's fine to enable the test for the development process and to use it when working on this code but until I can reduce the burden of it we should avoid enabling it by default, even if all the issues are fixed.

@ashleypittman ok, I will turn it back off for now.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mchaarawi reviewed (commented):
the name of the PR and ticket number are now wrong, right?

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mchaarawi reviewed (changes requested):
should probably just create a new ticket and change the title

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/24/display/redirect

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/24/display/redirect

dsikich commented:
@mchaarawi it looks like PR #6333 broke how fs copy and cont clone were using duns_resolve_path. Should passing in a daos://$pool no longer work? It only works now if I generate a cont UUID first now. 

```bash
[dsikich@boro-54 ftest]$ daos filesystem copy --src ~/daos-mpifileutils/dir_small --dst daos://$pool
DEBUG 01:05:39.270371 mgmt_system.go:48: MgmtSvc.GetAttachInfo dispatch, req:{state:{NoUnkeyedLiterals:{} DoNotCompare:[] DoNotCopy:[] atomicMessageInfo:0xc000297df8} sizeCache:0 unknownFields:[] Sys:daos_server AllRanks:true}
DEBUG 01:05:39.270945 mgmt_system.go:79: MgmtSvc.GetAttachInfo dispatch, resp:{state:{NoUnkeyedLiterals:{} DoNotCompare:[] DoNotCopy:[] atomicMessageInfo:<nil>} sizeCache:0 unknownFields:[] Status:0 RankUris:[uri:"ofi+sockets://192.168.1.54:31416"] MsRanks:[0] ClientNetHint:provider:"ofi+sockets" net_dev_class:32 srv_srx_set:-1} len(RankUris):1
ERROR: : failed to parse destination path: Invalid parameters (-1003)
ERROR: daos: failed to copy /home/dsikich/daos-mpifileutils/dir_small -> daos://b3fec8ef-5090-4edb-ac9c-3fed5a23fdc0: DER_INVAL(-1003): Invalid parameters
[dsikich@boro-54 ftest]$ uuidgen
29e5eb22-d871-4d40-b850-5cfafea45224
[dsikich@boro-54 ftest]$ daos filesystem copy --src ~/daos-mpifileutils/dir_small --dst daos://$pool/29e5eb22-d871-4d40-b850-5cfafea45224
DEBUG 01:08:26.625616 mgmt_drpc.go:102: handling GetPoolSvcReq: uuid:"b3fec8ef-5090-4edb-ac9c-3fed5a23fdc0"
DEBUG 01:08:26.625882 mgmt_drpc.go:115: GetPoolSvcResp: svcreps:0
Successfully created container 29e5eb22-d871-4d40-b850-5cfafea45224
Successfully copied to DAOS: /
    Directories: 6
    Files:       3
```

mchaarawi commented:
> @mchaarawi it looks like PR #6333 broke how fs copy and cont clone were using duns_resolve_path. Should passing in a daos://$pool no longer work? It only works now if I generate a cont UUID first now.
> 
> ```shell
> [dsikich@boro-54 ftest]$ daos filesystem copy --src ~/daos-mpifileutils/dir_small --dst daos://$pool
> DEBUG 01:05:39.270371 mgmt_system.go:48: MgmtSvc.GetAttachInfo dispatch, req:{state:{NoUnkeyedLiterals:{} DoNotCompare:[] DoNotCopy:[] atomicMessageInfo:0xc000297df8} sizeCache:0 unknownFields:[] Sys:daos_server AllRanks:true}
> DEBUG 01:05:39.270945 mgmt_system.go:79: MgmtSvc.GetAttachInfo dispatch, resp:{state:{NoUnkeyedLiterals:{} DoNotCompare:[] DoNotCopy:[] atomicMessageInfo:<nil>} sizeCache:0 unknownFields:[] Status:0 RankUris:[uri:"ofi+sockets://192.168.1.54:31416"] MsRanks:[0] ClientNetHint:provider:"ofi+sockets" net_dev_class:32 srv_srx_set:-1} len(RankUris):1
> ERROR: : failed to parse destination path: Invalid parameters (-1003)
> ERROR: daos: failed to copy /home/dsikich/daos-mpifileutils/dir_small -> daos://b3fec8ef-5090-4edb-ac9c-3fed5a23fdc0: DER_INVAL(-1003): Invalid parameters
> [dsikich@boro-54 ftest]$ uuidgen
> 29e5eb22-d871-4d40-b850-5cfafea45224
> [dsikich@boro-54 ftest]$ daos filesystem copy --src ~/daos-mpifileutils/dir_small --dst daos://$pool/29e5eb22-d871-4d40-b850-5cfafea45224
> DEBUG 01:08:26.625616 mgmt_drpc.go:102: handling GetPoolSvcReq: uuid:"b3fec8ef-5090-4edb-ac9c-3fed5a23fdc0"
> DEBUG 01:08:26.625882 mgmt_drpc.go:115: GetPoolSvcResp: svcreps:0
> Successfully created container 29e5eb22-d871-4d40-b850-5cfafea45224
> Successfully copied to DAOS: /
>     Directories: 6
>     Files:       3
> ```

i pushed a fix for now.
but the usage is still wrong and needs to be updated to take labels into account. This can be done in a different PR, but should be required for 2.0

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/26/display/redirect

daosbuild1 commented:
Test stage Functional Hardware Small completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/26/display/redirect

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/26/display/redirect

daosbuild1 commented:
Test stage Functional Hardware Small completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/27/display/redirect

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6353/28/display/redirect

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -3607,30 +3316,27 @@ cont_clone_hdlr(struct cmd_args_s *ap)
-			"Unable to allocate memory for dest path.");
+		rc = -DER_NOMEM;
+		DH_PERROR_DER(ap, rc, "Unable to allocate memory for destination path");
 		D_GOTO(out, rc = -DER_NOMEM);
```
You're setting rc twice here.  Minor ;)
```suggestion
		D_GOTO(out, rc);
```

ashleypittman commented on utils/node_local_test.py:
```diff
@@ -3436,8 +3436,8 @@ def run(wf, args):
         if fi_test:
-#            fatal_errors.add_result(test_alloc_fail_copy(server, conf,
-#                                                         wf_client))
+            #fatal_errors.add_result(test_alloc_fail_copy(server, conf,
+            #                                             wf_client))
```
Whitespace only change.

dsikich replied:
@ashleypittman Since there are minor I can address these in a follow on PR I have for this file. Some of the parsing for all the datamover tools broke after PR #6333 was merged and there is a fix in here for it.

ashleypittman replied:
I only mention it because it's going to conflict with one of mine.
